test/gateway/eip20_gateway/progress_unstake.js

```diff
@@ -0,0 +1,535 @@
+    baseToken = await MockToken.new({ from: accounts[0] });
+
+    let organizationAddress = accounts[3];
+    let coreAddress = accounts[5];
+    let burnerAddress = NullAddress;
+
+    gateway = await Gateway.new(
+      mockToken.address,
+      baseToken.address,
+      coreAddress,
+      bountyAmount,
+      organizationAddress,
+      burnerAddress,
+    );
+
+    unstakeRequest = {
+      beneficiary: accounts[6],
+      amount: new BN(100000000),
+    };
+
+    let hashLockObj = Utils.generateHashLock();
+
+    unstakeMessage = {
+      intentHash: web3.utils.sha3("dummy"),
+      nonce: new BN(1),
+      gasPrice: new BN(100),
+      gasLimit: new BN(1000000),
+      unstakeAccount: accounts[8],
+      hashLock: hashLockObj.l,
+      unlockSecret: hashLockObj.s
+    };
+
+    await setMessage();
+
+    stakeVaultAddress = await gateway.stakeVault.call();
+
+    await mockToken.transfer(
+      stakeVaultAddress,
+      new BN(500000000),
+      { from: accounts[0] }
+    );
+
+  });
+
+  it('should fail when message hash is zero', async function () {
+
+    let messageHash = ZeroBytes;
+
+    await Utils.expectRevert(
+      gateway.progressUnstake(
+        messageHash,
+        unstakeMessage.unlockSecret,
+      ),
+      'Message hash must not be zero.',
+    );
+
+  });
+
+  it('should fail when unlock secret is incorrect', async function () {
+
+    let unlockSecret = ZeroBytes;
+
+    await Utils.expectRevert(
+      gateway.progressUnstake(
+        unstakeMessage.messageHash,
+        unlockSecret,
+      ),
+      'Invalid unlock secret.',
+    );
+
+  });
+
+  it('should fail when unstake message is undeclared', async function () {
+
+    await Utils.expectRevert(
+      gateway.progressUnstake(
+        unstakeMessage.messageHash,
+        unstakeMessage.unlockSecret,
+      ),
+      'Message on target status must be Declared.',
+    );
+
+  });
+
+  it('should fail when unstake message is already progressed', async function () {
+
+    await gateway.setInboxStatus(
+      unstakeMessage.messageHash,
+      MessageStatusEnum.Declared,
+    );
+
+    await gateway.progressUnstake(
+      unstakeMessage.messageHash,
+      unstakeMessage.unlockSecret,
+    );
+
+    await Utils.expectRevert(
+      gateway.progressUnstake(
+        unstakeMessage.messageHash,
+        unstakeMessage.unlockSecret,
+      ),
+      'Message on target status must be Declared.',
+    );
+
+  });
+
+  it('should fail for revoked redeem(unstake) message', async function () {
+
+    await gateway.setInboxStatus(
+      unstakeMessage.messageHash,
+      MessageStatusEnum.Revoked,
+    );
+
+    await Utils.expectRevert(
+      gateway.progressUnstake(
+        unstakeMessage.messageHash,
+        unstakeMessage.unlockSecret,
+      ),
+      'Message on target status must be Declared.',
+    );
+
+  });
+
+  it('should fail when redeem(unstake) message status is revocation declared',
+    async function () {
```

Comment:
To me, tests are documentation of feature. This test is misleading, as this is not possible.

Reply:
Deleted it